### PR TITLE
Refactored new FakeUpstream.

### PR DESCRIPTION
**Pull request by mk46**
Signed-off-by: Manish Kumar <manish.kumar1@india.nec.com>

 
Commit Message: Refactored Fake Upstream.
Additional Description: Refactoring fake upstream instantiation. 
Risk Level:
Testing:
Docs Changes:
Release Notes:
includes partial fix to #11109

### CODE REVIEW

**Comment by alyssawilk** [comment]
I'd have to check with @jmarantz but I think this doesn't reduce the clutter he was hoping to.

I think what he had in mind was a helper function in the integration test
createFakeUpstream() which would return a new fake upstream without all of the arguments we used to pass into std::make_unique<FakeUpstream>

Make sense?

**Comment by jmarantz** [comment]
Agreed -- what Alyssa said is what I was suggesting in the bug.

**Comment by mk46** [comment]
@alyssawilk Would you mind take a look? Thanks!

File: test/integration/base_integration_test.h [hunk 1]
```diff
@@ -289,6 +289,32 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
...
+  }
+  // Creates a fake upstream bound to the specified |address|.
+  std::unique_ptr<FakeUpstream>
+  createFakeUpstream(const Network::Address::InstanceConstSharedPtr& address,
+                     FakeHttpConnection::Type type, bool enable_half_close = false,
+                     bool udp_fake_upstream = false) {
+    return std::make_unique<FakeUpstream>(address, type, timeSystem(), enable_half_close,
+                                          udp_fake_upstream);
+  }
+  // Creates a fake upstream bound to INADDR_ANY and the specified |port|.
```
Comment (alyssawilk) [inline test/integration/base_integration_test.h:306]: Mind updating the comment - there is no specified port :-)

**Review by alyssawilk — COMMENTED** [review]
Awesome - I think this is the cleanup @jmarantz was looking for.
A couple of questions/thoughts below

File: test/integration/base_integration_test.h (hunk 1 above)
Comment (mk46) [reply test/integration/base_integration_test.h]: Done.

File: test/integration/base_integration_test.h [hunk 2]
```diff
@@ -289,6 +289,32 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
...
 
+  // Helper to create FakeUpstream.
+  // Creates a fake upstream bound to the specified unix domain socket path.
+  std::unique_ptr<FakeUpstream> createFakeUpstream(const std::string& uds_path,
+                                                   FakeHttpConnection::Type type) {
+    return std::make_unique<FakeUpstream>(uds_path, type, timeSystem());
+  }
+  // Creates a fake upstream bound to the specified |address|.
+  std::unique_ptr<FakeUpstream>
+  createFakeUpstream(const Network::Address::InstanceConstSharedPtr& address,
```
[the remark this answers is not in the document]
Comment (mk46) [reply test/integration/base_integration_test.h:300]: enable_half_close_  is false for all, so removed from args.

**Review by jmarantz — COMMENTED** [review]
This is a nice improvement, thank you!

One other possible improvement: almost (not all) of the call-sites to createFakeUpstream are passed directly to fake_upstreams_.emplace_back_. You could make things even leaner by replacing those with a new abstraction addFakeUpstream(...). You need to leave the createFakeUpstream entry point as there are a couple of places that its return value is assigned to a local.

But this is already a great improvement!

File: test/integration/base_integration_test.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/integration/base_integration_test.h:300]: Nice cleanup, thanks!  Did you happen check on type == upstream_protocol_ as well?  Not sure if it didn't work out or you missed that bit.

[the remark this answers is not in the document]
Comment (mk46) [reply test/integration/base_integration_test.h:300]: upstream_protocol_ is set to "FakeHttpConnection::Type::HTTP1" inside base_integration_test and enable_half_close_ is independant of whether type is HTTP1(upstream_protocol_) or HTTP2. Does it make sense to checking type?

**Comment by mk46** [comment]
@jmarantz Would you mind to take a look? Thanks!

File: test/integration/base_integration_test.h [hunk 3]
```diff
@@ -289,6 +289,36 @@ class BaseIntegrationTest : protected Logger::Loggable<Logger::Id::testing> {
...
+    return std::make_unique<FakeUpstream>(0, type, version_, timeSystem(), enable_half_close);
+  }
+  std::unique_ptr<FakeUpstream>
+  createFakeUpstream(Network::TransportSocketFactoryPtr&& transport_socket_factory,
+                     FakeHttpConnection::Type type) {
+    return std::make_unique<FakeUpstream>(std::move(transport_socket_factory), 0, type, version_,
+                                          timeSystem());
+  }
+  // Helper to add FakeUpstream.
+  void addFakeUpstream(std::unique_ptr<FakeUpstream> fake_upstream) {
```
Comment (jmarantz) [inline test/integration/base_integration_test.h:318]: I was thinking `addFakeUpstream` would call createFakeUpstream itself, which would further factor out a bunch of repeated patterns.

Comment (mk46) [reply test/integration/base_integration_test.h:318]: Thanks, will do.

**Comment by mk46** [comment]
@jmarantz Could you take a look? Thanks!

File: test/integration/base_integration_test.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/integration/base_integration_test.h:300]: sorry, my point is that I believe in this function that type will always be the same as upstream_protocol_

If you could verify that by adding a temporary ASSERT(upstream_protocol_ == type) and running the integration tests, you could remove type as an argument as upstream_protocol_ can be used instead.

[the remark this answers is not in the document]
Comment (mk46) [reply test/integration/base_integration_test.h:300]: I have checked, the type is not the same as upstream_protocol_.

[the remark this answers is not in the document]
Comment (alyssawilk) [reply test/integration/base_integration_test.h:300]: gotcha, thanks!

**Review by jmarantz — APPROVED** [review]

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**